### DISCUSSION TITLE
Rainfall induced stability in porous flow using tensormechanics module

### DISCUSSION THREAD

**Discussion by delpierrena** [idaholab/moose#24087] [Q&A Modules: Solid mechanics] [answered]
Hello, 

I'm trying to simulate slope instabilities due to rainfalls in the module PorousFlow. To do so, I need to simulate the infiltration of water with the Richards equation and then use the TensorMechanicsPlasticMohrCoulomb plastic model that depends on the degree of saturation. I will have heterogeneous plastic properties (cohesion and friction angle) that vary at each Gauss points. 

But for the moment, the parameters of the Mohr Coulomb model (cohesion and friction angle) are defined through an User Object in the pattern of TensorMechanicsHardeningModel which only depends on the accumulated plastic strain and the stress level, assuming the same yield function parameters in the whole domain.  

Is there a way to define the value of cohesion and friction at the nodes or cells center and not as constant for the whole material ? 
Or do I need to change the implementation more in depth ? 

Thank you ! 
Nathan

**Comment by WilkAndy** [comment, marked as answer]
Unfortunately, it is not possible to define the value of cohesion and friction at the nodes or cells center, only by Material.  I think you're going to have to do C++ work, boohoo.

**Reply by delpierrena** [reply]
Hello, 
In order to use the TensorMechanicsPlasticMohrCoulomb UO with heterogeneous mechanical properties, I am trying to build a new UO based on the TensorMechanicsHardeningModel. My initial idea was to give a FunctionAux as a required parameter : 
``` 
InputParameters
TensorMechanicsHardeningHeterogeneous::validParams()
{
  InputParameters params = TensorMechanicsHardeningModel::validParams();
  params.addRequiredParam<FunctionName>(    "value",
                                            "The value of the parameter for all internal parameter.  "
                                            "The value will decrease linearly with time.");
  params.addParam<bool>("convert_to_radians",
                        false,
                        "If true, the value you entered will be "
                        "multiplied by Pi/180 before passing to the "
                        "Plasticity algorithms");
  params.addClassDescription(
      "No hardening - the parameter only depends on time");
  return params;
}
``` 
Then I build my specific UO as follows :

``` 
 TensorMechanicsHardeningHeterogeneous::TensorMechanicsHardeningHeterogeneous(
    const InputParameters & parameters)
  : TensorMechanicsHardeningModel(parameters),
    _val(parameters)
{
}
``` 
And this is where my issue starts because I dont understand how to specifically access the value taken at the current quadrature point :  

``` 
Real TensorMechanicsHardeningHeterogeneous::value(Real /*intnl*/) const { return _val; }

Real TensorMechanicsHardeningHeterogeneous::derivative(Real /*intnl*/) const { return 0.0; }

std::string
TensorMechanicsHardeningHeterogeneous::modelName() const
{
  return "ConstantTime";
}
``` 
Of course the value function throws back an error when compiling 
``` 
 error: cannot convert 'const FunctionAux' to 'libMesh::Real' {aka 'double'} in return
   39 | Real TensorMechanicsHardeningHeterogeneous::value(Real /*intnl*/) const { return _val; }
``` 

Do you think of any way to do this  ? Or I am going in the wrong direction? 

Thank you very much ! 
Nathan

**Reply by GiudGiud** [reply]
Hello

this isn’t quite how to retrieve a function

please use getFunction in the constructor s initialization list (see GenericFunctionMaterial for an example)

and same when you want to use a function to get a value you don’t return the function itself (or it’s name) you use the function object and call the value(time, …) routine on it
